File: runtime/audit_guard.py
"""Fail-closed workspace snapshots for independent Auditor episodes."""

from __future__ import annotations

from dataclasses import dataclass
import hashlib
import os
from pathlib import Path
from typing import Mapping

from .core.contracts import ContractError


_EXCLUDED_ROOTS = {".git", ".trellis", ".expert-team", "__pycache__"}


@dataclass(frozen=True)
class WorkspaceEntry:
    kind: str
    size: int
    digest: str


@dataclass(frozen=True)
class WorkspaceSnapshot:
    root: Path
    entries: Mapping[str, WorkspaceEntry]
    errors: tuple[str, ...]

    @property
    def complete(self) -> bool:
        return not self.errors


@dataclass(frozen=True)
class WorkspaceDiff:
    added: tuple[str, ...]
    deleted: tuple[str, ...]
    changed: tuple[str, ...]
    type_changed: tuple[str, ...]
    errors: tuple[str, ...]

    @property
    def clean(self) -> bool:
        return not (self.added or self.deleted or self.changed or self.type_changed or self.errors)


def snapshot_workspace(
    root: Path,
    *,
    max_files: int = 20_000,
    max_hash_bytes: int = 2_000_000,
) -> WorkspaceSnapshot:
    resolved = root.resolve()
    if not resolved.is_dir():
        raise ContractError("workspace snapshot root must be an existing directory")
    if max_files < 1 or max_hash_bytes < 1:
        raise ContractError("workspace snapshot limits must be positive")
    entries: dict[str, WorkspaceEntry] = {}
    errors: list[str] = []
    paths: list[Path] = []

    def on_error(error: OSError) -> None:
        errors.append(str(error))

    for current, directories, files in os.walk(resolved, topdown=True, followlinks=False, onerror=on_error):
        current_path = Path(current)
        directories[:] = sorted(name for name in directories if name not in _EXCLUDED_ROOTS)
        for name in directories:
            paths.append(current_path / name)
        for name in sorted(files):
            if name not in _EXCLUDED_ROOTS:
                paths.append(current_path / name)
    for path in paths:
        relative = path.relative_to(resolved)
        if len(entries) >= max_files:
            errors.append(f"file limit exceeded: {max_files}")
            break
        key = relative.as_posix()
        try:
            if path.is_symlink():
                target = os.readlink(path)
                entries[key] = WorkspaceEntry("symlink", len(target), hashlib.sha256(target.encode()).hexdigest())
            elif path.is_dir():
                entries[key] = WorkspaceEntry("directory", 0, "")
            elif path.is_file():
                stat = path.stat()
                if stat.st_size > max_hash_bytes:
                    errors.append(f"file exceeds hash limit: {key}")
                    continue
                digest = hashlib.sha256(path.read_bytes()).hexdigest()
                entries[key] = WorkspaceEntry("file", stat.st_size, digest)
            else:
                entries[key] = WorkspaceEntry("other", 0, "")
        except OSError as error:
            errors.append(f"{key}: {error}")
    return WorkspaceSnapshot(resolved, entries, tuple(errors))


def diff_workspace(before: WorkspaceSnapshot, after: WorkspaceSnapshot) -> WorkspaceDiff:
    if before.root != after.root:
        raise ContractError("workspace snapshots have different roots")
    before_keys = set(before.entries)
    after_keys = set(after.entries)
    added = sorted(after_keys - before_keys)
    deleted = sorted(before_keys - after_keys)
    changed: list[str] = []
    type_changed: list[str] = []
    for key in sorted(before_keys & after_keys):
        left = before.entries[key]
        right = after.entries[key]
        if left.kind != right.kind:
            type_changed.append(key)
        elif left != right:
            changed.append(key)
    return WorkspaceDiff(
        tuple(added),
        tuple(deleted),
        tuple(changed),
        tuple(type_changed),
        (*before.errors, *after.errors),
    )

File: src/runtime/supervisor/audit-guard.ts
import { createHash } from "node:crypto";
import { lstat, opendir, readFile, readlink, realpath } from "node:fs/promises";
import { relative, resolve, sep } from "node:path";

import { ContractError } from "../core/contracts.js";

const excludedRoots = new Set([".git", ".trellis", ".expert-team", "__pycache__", "node_modules", "dist"]);
export interface WorkspaceEntry { kind: "file" | "directory" | "symlink" | "other"; size: number; digest: string }
export interface WorkspaceSnapshot { root: string; entries: Record<string, WorkspaceEntry>; errors: string[] }
export interface WorkspaceDiff { added: string[]; deleted: string[]; changed: string[]; typeChanged: string[]; errors: string[]; clean: boolean }

export async function snapshotWorkspace(root: string, options: { maxFiles?: number; maxHashBytes?: number } = {}): Promise<WorkspaceSnapshot> {
  const maxFiles = options.maxFiles ?? 20_000;
  const maxHashBytes = options.maxHashBytes ?? 2_000_000;
  if (maxFiles < 1 || maxHashBytes < 1) throw new ContractError("workspace snapshot limits must be positive");
  let resolved: string;
  try { resolved = await realpath(root); } catch (error) { throw new ContractError("workspace snapshot root must be an existing directory", { cause: error }); }
  const entries: Record<string, WorkspaceEntry> = {};
  const errors: string[] = [];
  const walk = async (directory: string): Promise<void> => {
    let handle;
    try { handle = await opendir(directory); } catch (error) { errors.push(String(error)); return; }
    const children = [];
    for await (const child of handle) children.push(child);
    children.sort((a, b) => a.name.localeCompare(b.name));
    for (const child of children) {
      if (excludedRoots.has(child.name)) continue;
      if (Object.keys(entries).length >= maxFiles) { errors.push(`file limit exceeded: ${maxFiles}`); return; }
      const path = resolve(directory, child.name);
      const key = relative(resolved, path).split(sep).join("/");
      try {
        const info = await lstat(path, { bigint: false });
        if (child.isSymbolicLink()) {
          const target = await readlink(path);
          entries[key] = { kind:"symlink", size:target.length, digest:createHash("sha256").update(target).digest("hex") };
        } else if (child.isDirectory()) {
          entries[key] = { kind:"directory", size:0, digest:"" };
          await walk(path);
        } else if (child.isFile()) {
          if (info.size > maxHashBytes) { errors.push(`file exceeds hash limit: ${key}`); continue; }
          entries[key] = { kind:"file", size:info.size, digest:createHash("sha256").update(await readFile(path)).digest("hex") };
        } else entries[key] = { kind:"other", size:0, digest:"" };
      } catch (error) { errors.push(`${key}: ${String(error)}`); }
    }
  };
  await walk(resolved);
  return { root:resolved, entries, errors };
}

export function diffWorkspace(before: WorkspaceSnapshot, after: WorkspaceSnapshot): WorkspaceDiff {
  if (before.root !== after.root) throw new ContractError("workspace snapshots have different roots");
  const beforeKeys = new Set(Object.keys(before.entries));
  const afterKeys = new Set(Object.keys(after.entries));
  const added = [...afterKeys].filter((key) => !beforeKeys.has(key)).sort();
  const deleted = [...beforeKeys].filter((key) => !afterKeys.has(key)).sort();
  const changed: string[] = [];
  const typeChanged: string[] = [];
  for (const key of [...beforeKeys].filter((value) => afterKeys.has(value)).sort()) {
    const left = before.entries[key]!; const right = after.entries[key]!;
    if (left?.kind !== right?.kind) typeChanged.push(key);
    else if (left?.size !== right?.size || left.digest !== right.digest) changed.push(key);
  }
  const errors = [...before.errors, ...after.errors];
  return { added, deleted, changed, typeChanged, errors, clean:added.length + deleted.length + changed.length + typeChanged.length + errors.length === 0 };
}
